src/Agent.Worker/TaskRunner.cs

```diff
@@ -370,6 +350,48 @@ public async Task RunAsync()
             }
         }
 
+        public HandlerData GetHandlerData(IExecutionContext ExecutionContext, ExecutionData currentExecution, PlatformUtil.OS hostOS)
+        {
+            if (currentExecution == null)
+            {
+                return null;
+            }
+
+            if ((currentExecution.All.Any(x => x is PowerShell3HandlerData)) &&
+                (currentExecution.All.Any(x => x is PowerShellHandlerData && x.Platforms != null && x.Platforms.Contains("windows", StringComparer.OrdinalIgnoreCase))))
+            {
+                // When task contains both PS and PS3 implementations, we will always prefer PS3 over PS regardless of the platform pinning.
+                Trace.Info("Ignore platform pinning for legacy PowerShell execution handler.");
+                var legacyPShandler = currentExecution.All.Where(x => x is PowerShellHandlerData).FirstOrDefault();
+                legacyPShandler.Platforms = null;
+            }
+
+            var targetOS = hostOS;
+
+            var stepTarget = ExecutionContext.StepTarget();
+            var preferPowershellHandler = true;
+            var preferPowershellHandlerOnContainers = ExecutionContext.Variables.GetBoolean("agent.preferPowerShellOnContainers")
+                ?? StringUtil.ConvertToBoolean(System.Environment.GetEnvironmentVariable("AGENT_PREFER_POWERSHELL_ON_CONTAINERS"), false);
+            if (!preferPowershellHandlerOnContainers && stepTarget != null)
+            {
+                targetOS = stepTarget.ExecutionOS;
+                if (stepTarget is ContainerInfo)
+                {
+                    if ((currentExecution.All.Any(x => x is PowerShell3HandlerData)) &&
```

Comment:
Nit - we probably don't need to check for `x => x is PowerShell3HandlerData` here